src/target/xtensa/xtensa.c

```diff
@@ -975,58 +1059,73 @@ int xtensa_fetch_all_regs(struct target *target)
 		for (unsigned int i = 0; i < 16; i++) {
 			if (i + j < xtensa->core_config->aregs_num) {
 				xtensa_queue_exec_ins(xtensa,
-					XT_INS_WSR(XT_SR_DDR, xtensa_regs[XT_REG_IDX_AR0 + i].reg_num));
-				xtensa_queue_dbg_reg_read(xtensa, NARADR_DDR, regvals[XT_REG_IDX_AR0 + i + j]);
+					XT_INS_WSR(xtensa, XT_SR_DDR, xtensa_regs[XT_REG_IDX_AR0 + i].reg_num));
+				xtensa_queue_dbg_reg_read(xtensa, NARADR_DDR, 
+					(uint8_t *)&regvals[XT_REG_IDX_AR0 + i + j]);
 				if (debug_dsrs)
-					xtensa_queue_dbg_reg_read(xtensa, NARADR_DSR, dsrs[XT_REG_IDX_AR0 + i + j]);
+					xtensa_queue_dbg_reg_read(xtensa, NARADR_DSR, 
+						(uint8_t *)&dsrs[XT_REG_IDX_AR0 + i + j]);
 			}
 		}
 		if (xtensa->core_config->windowed) {
 			/* Now rotate the window so we'll see the next 16 registers. The final rotate
 			 * will wraparound, */
 			/* leaving us in the state we were. */
-			xtensa_queue_exec_ins(xtensa, XT_INS_ROTW(4));
+			xtensa_queue_exec_ins(xtensa, XT_INS_ROTW(xtensa, 4));
 		}
 	}
+	xtensa_window_state_restore(target, woe);
+
 	if (xtensa->core_config->coproc) {
 		/* As the very first thing after AREGS, go grab the CPENABLE registers. It indicates
 		 * if we can also grab the FP */
 		/* (and theoretically other coprocessor) registers, or if this is a bad thing to do.*/
-		xtensa_queue_exec_ins(xtensa, XT_INS_RSR(xtensa_regs[XT_REG_IDX_CPENABLE].reg_num, XT_REG_A3));
-		xtensa_queue_exec_ins(xtensa, XT_INS_WSR(XT_SR_DDR, XT_REG_A3));
-		xtensa_queue_dbg_reg_read(xtensa, NARADR_DDR, regvals[XT_REG_IDX_CPENABLE]);
+		xtensa_queue_exec_ins(xtensa, XT_INS_RSR(xtensa, xtensa_regs[XT_REG_IDX_CPENABLE].reg_num, XT_REG_A3));
+		xtensa_queue_exec_ins(xtensa, XT_INS_WSR(xtensa, XT_SR_DDR, XT_REG_A3));
+		xtensa_queue_dbg_reg_read(xtensa, NARADR_DDR, (uint8_t *)&regvals[XT_REG_IDX_CPENABLE]);
 	}
-	int res = jtag_execute_queue();
+	res = jtag_execute_queue();
 	if (res != ERROR_OK) {
 		LOG_ERROR("Failed to read ARs (%d)!", res);
 		return res;
 	}
 	xtensa_core_status_check(target);
 
+	a3 = buf_get_u32(a3_buf, 0, 32);
+
 	if (xtensa->core_config->coproc)
-		cpenable = buf_get_u32(regvals[XT_REG_IDX_CPENABLE], 0, 32);
+		cpenable = buf_get_u32((uint8_t *)&regvals[XT_REG_IDX_CPENABLE], 0, 32);
```

Comment:
host endian issue

Reply:
Actually I think this should work--entries in regvals[] and dsrs[] are always treated as uint8_t buffers, and are used to queue debug reads or passed to buf_get_u32().  The definition is what's confusing, since it's not really a uint32_t (but is defined that way for the convenience of dereferencing).  

The problem is that this is now a dynamically-sized list instead of a statically-defined array.  It might be clearer to define these as lists of a union element (xtensa_reg_val_t || uint8_t[4]), then use the uint8_t part of the union throughout the function.

I don't see any objection to unions in the OOCD C style guidelines.  Do you think this is a better approach?

Reply:
Yeah, this looks better to me too.  Thanks.